### PR TITLE
Add build manifest

### PR DESCRIPTION
**Pull request by timneutkens**
This PR does 2 things:

1. Makes main.js / manifest.js use chunkhash in production
2. Add buildManifest as parameter for `renderPage` which allows users extending the webpack config to use chunkhash for files too. One example of this is that we can modify next-css to use chunkhash in production mode.

### CODE REVIEW

**Comment by timneutkens** [comment]
Custom document when adding css will look like:

```
// _document is only rendered on the server side and not on the client side
// Event handlers like onClick can't be added to this file

// ./pages/_document.js
import Document, { Head, Main, NextScript } from 'next/document'
import flush from 'styled-jsx/server'

export default class MyDocument extends Document {
  render() {
    const {buildManifest, pathname} = this.props
    const {css} = buildManifest
    return (
      <html>
        <Head>
          <style>{`body { margin: 0 } /* custom! */`}</style>
          {css.map((file) => {
            return <link rel="stylesheet" href={`/_next/${file}`} key={file} />
          })}
          {}
        </Head>
        <body className="custom_class">
          {this.props.customValue}
          <Main />
          <NextScript />
        </body>
      </html>
    )
  }
}
```

**Comment by timneutkens** [comment]
Build manifest
```
{
  "pages": {
    "/_error": {
      "pageFile": "bundles/pages/_error.js",
      "files": []
    },
    "/": {
      "pageFile": "bundles/pages/index.js",
      "files": [
        "static/style.css"
      ]
    }
  },
  "css": [
    "static/style.css"
  ],
  "main.js": [
    "static/commons/main.js"
  ],
  "manifest.js": [
    "static/commons/manifest.js"
  ]
}
```

File: server/build/plugins/build-manifest-plugin.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          if (/\.map$/.test(file)) {
+            continue
+          }
+
+          if (/\.hot-update\.js$/.test(file)) {
+            continue
+          }
+
+          if (MATCH_ROUTE_NAME.exec(file)) {
+            pageFile = file
```
Comment (arunoda) [inline server/build/plugins/build-manifest-plugin.js]: Why store this inside a loop?
Isn't that the previous value could get replaced when we are going to use this?
What's the purpose?

File: server/document.js [hunk 2]
```diff
@@ -125,31 +126,32 @@ export class NextScript extends Component {
...
   getChunkScript (filename, additionalProps = {}) {
-    const { __NEXT_DATA__ } = this.context._documentProps
+    const { __NEXT_DATA__, buildManifest } = this.context._documentProps
     let { assetPrefix, buildId } = __NEXT_DATA__
-    const hash = buildId
 
-    return (
+    const files = buildManifest[filename]
+
+    return files.map((file) => (
```
Comment (arunoda) [inline server/document.js]: Why we are having multiple files?

**Comment by barrysteyn** [comment]
@timneutkens 
This change is failing for me, however it seems like in the manifest, the webpack compiler is giving an incorrect chunk name.

For example, in my `next.config.js`, I do this:

```
const withHashedOutput = (nextConfig = {}) => {
  return Object.assign({}, nextConfig, {
    webpack(config, options) {
      if (!options.defaultLoaders) {
        throw new Error(
          'This plugin is not compatible with Next.js versions below 5.0.0 https://err.sh/next-plugins/upgrade'
        );
      }

      config.output.filename = "[name][id]";

      if (typeof nextConfig.webpack === 'function') {
        return nextConfig.webpack(config, options);
      }

      return config;
    }
  });
};

const config = withHashedOutput({});

module.exports = config;
```

So I am just changing the file name by putting it's id. However, the manifest gets the incorrect id number - this is because webpack seems to be returning incorrect information. I am trying this out on `6.0.0-canary.7`

Why is this happening? You can easily repro whats happening by using the above - you will see, it won't work.

**Pull request merged**